### PR TITLE
Introduce "fold" as an alternative to "reduce"

### PR DESCRIPTION
**Pull request by atilaneves**
I took the easy way out here; I have no idea why `reduce` throws when passed an empty range and I don't think `fold` should do it either. However, since it's a lot harder to make `fold` `nothrow` I'll leave it as it is right now and maybe submit a PR later to make it so. IMHO, this should compile and pass:

    @safe pure nothrow unittest
    {
        int[] arr;
        assert(arr.fold!((a, b) => a + b) == 0);
        assert(arr.fold!(min, max) == tuple(0, 0));
    }

i.e. return `Unqual!(ElementType!R).init` when passed in an empty range. It gets more complicated since `reduce` also supports `opApply`.

### CODE REVIEW

**Comment by MetaLang** [comment]
> I have no idea why reduce throws when passed an empty range
I believe this is just historical baggage. Today it would be written as `assert(!range.empty)` or just leave it up to the range to assert when front/popFront are called and empty = true.

File: std/algorithm/iteration.d [hunk 1]
```diff
@@ -39,6 +39,8 @@ $(T2 permutations,
         Lazily computes all permutations using Heap's algorithm.)
 $(T2 reduce,
         $(D reduce!((a, b) => a + b)([1, 2, 3, 4])) returns $(D 10).)
+$(T2 fold,
```
Comment (dcarp) [inline std/algorithm/iteration.d]: Don't break the lexical order.

**Comment by dcarp** [comment]
Changelog entry is missing.

**Comment by dcarp** [comment]
For a smoother deprecation path, it will be better to use `fold` in the original implementation and create `reduce` as an wrapper of `fold`.
Initially the unit tests can still use `reduce`. This will assure that the wrapper implementation works in all the tested cases.

**Comment by dcarp** [comment]
The reference in `std.algorithm.package.d`s ddoc is missing.

**Comment by atilaneves** [comment]
I don't think it's a good idea to use fold as the "real" implementation. It's a lot more work for little to no benefit, and contrary to what was discussed [here](http://forum.dlang.org/post/n8taih$6dg$2@digitalmars.com).

File: std/algorithm/iteration.d [hunk 2]
```diff
@@ -2864,6 +2866,77 @@ private template ReduceSeedType(E)
...
+    assert(arr.fold!((a, b) => a + b)(6) == 21);
+
+    import std.algorithm.comparison: min, max;
+    import std.typecons: tuple;
+
+    // Compute minimum and maximum at the same time
+    assert(arr.fold!(min, max) == tuple(1, 5));
+
+    // Compute minimum and maximum at the same time with seeds
+    assert(arr.fold!(min, max)(0, 7) == tuple(0, 7));
```
Comment (quickfur) [inline std/algorithm/iteration.d]: What about adding an example of a UFCS chain, which is the primary motivation for introducing this function over `reduce` in the first place?

**Comment by quickfur** [comment]
Looks like a rebase is in order.

**Comment by atilaneves** [comment]
Added UFCS example and rebased.

File: std/algorithm/iteration.d [hunk 3]
```diff
@@ -2868,6 +2870,80 @@ private template ReduceSeedType(E)
...
+
+    // Can be used in a UFCS chain
+    assert(arr.map!(a => a +1).fold!((a, b) => a + b) == 20);
```
Comment (quickfur) [inline std/algorithm/iteration.d]: Nitpick: insert space after `+`, i.e., `+ 1` instead of `+1`.

**Comment by quickfur** [comment]
Also, please add `fold` to the ddoc index in `std/algorithm/package.d`.

**Comment by greenify** [comment]
Nice! I came by this issue from https://github.com/D-Programming-Language/phobos/pull/1955 and thought this will never make it into phobos. Thanks @atilaneves!

I would recommend to add some comments to `reduce` about `fold` - newbies like I am are otherwise confused how to call `reduce` in UCFS syntax with seeds ;-)

**Comment by atilaneves** [comment]
I'm on holiday right now, will address the most recent comments when I get back home

**Comment by atilaneves** [comment]
Comments addressed.

**Comment by quickfur** [comment]
Thanks! LGTM.

**Comment by quickfur** [comment]
ping random reviewers @JakobOvrum @andralex @yebblies @klickverbot 
This has been green for a while. Any other comments that need to be addressed, or can we merge this now?

File: std/algorithm/iteration.d [hunk 4]
```diff
@@ -2868,6 +2872,80 @@ private template ReduceSeedType(E)
     }
 }
 
+
+/++
+Implements the homonym function (also known as $(D accumulate), $(D
```
Comment (andralex) [inline std/algorithm/iteration.d]: Would be nice to use backticks instead of `$(D ...)` throughout (not a blocker).

**Comment by andralex** [comment]
This looks good. Will pull as is but please there is one follow-up task for @atilaneves: please make one more pass through all documentation and replace all uses of `reduce` with `fold`. Also deemphasize `reduce` to send people to `fold`. Essentially, even though fold is implemented in terms of reduce, the documentation and examples should go the other way around. Thanks!

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**